File: Lab3/GFF+LK.py
import cv2
import numpy as np

#cap = cv2.VideoCapture(0)
cap = cv2.VideoCapture("/home/mmlab/workspace/CVLaboratories/material/Video.mp4")

frame_index = 0
while cap.isOpened():
    #read video
    ret, frame = cap.read()

    #copy frame and convert to grayscale
    frame_copy = frame.copy()
    frame_gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # Select GFF Features
    if frame_index % 30 == 0:
        corners = cv2.goodFeaturesToTrack(frame_gray, maxCorners=100,
            qualityLevel=0.01,
            minDistance=10,
            blockSize=3,
            useHarrisDetector=False,
            k=0.04)
    else:
        #Track GFF features with Lucas-Kanade optical flow
        corners, status, err = cv2.calcOpticalFlowPyrLK(prev_frame, frame, prev_corners, None)

    #plot keypoints
    np_corners = np.int0(corners)
    for i,crn in enumerate(np_corners):
        x, y = crn.ravel()
        cv2.circle(frame_copy, (x, y), 3, np.array([i, 2*i, 255-i], float))

    #copy values for next iteration 
    prev_frame, prev_corners = frame.copy(), corners

    #plot results
    cv2.imshow('gff', frame_copy)
    if cv2.waitKey(1) & 0xFF == ord('q'): #close video is q is pressed
        break

    frame_index += 1

cap.release()
cv2.destroyAllWindows()

File: extra/C++/Lab 3 - Tracking/01 GFF + LK/GFF+LK.cpp
#include <opencv2/opencv.hpp>

using namespace cv;
using namespace std;

int main()
{

	Mat frame, prev_frame, frame_gray, copy;
    VideoCapture cap;

    int nFrames = 1000;
    int step = 100;
    int maxCorners = 200;

    /// Parameters for Shi-Tomasi algorithm
    vector<Point2f> corners, prev_corners;
    vector<uchar> status;
    vector<float> err;
    double qualityLevel = 0.01;
    double minDistance = 10;
    int blockSize = 3;
    bool useHarrisDetector = false;
    double k = 0.04;

    cap.open(0);
    if(!cap.isOpened())
    	return 0;

    namedWindow("Window",1);

    for(int i = 0; i<nFrames; i++){

    	cap >> frame;

    	copy = frame.clone();

    	cvtColor( frame, frame_gray, COLOR_BGR2GRAY );

    	if ( i % step == 0)
    	{
		//select features to track using GFF
    		goodFeaturesToTrack( frame_gray, corners, maxCorners, qualityLevel, minDistance, Mat(), blockSize, useHarrisDetector, k );

    	}
    	else
    	{
		//tracking
    		calcOpticalFlowPyrLK(prev_frame, frame, prev_corners, corners, status, err);

    	}

        int r = 4;
        for( int j = 0; j < corners.size(); j++ )
           {
        	circle( copy, corners[j], r, Scalar(5*j, 2*j, 255 - j), -1, 8, 0 );
           }

    	prev_frame = frame.clone();
    	prev_corners = corners;

    	imshow("Window", copy);
    	if(waitKey(10) >= 0) break;

    }


    return(0);
}

